### PR TITLE
theta integration

### PR DESCRIPTION
**Pull request by oksanaguba**
New branch/PR to reorg code compared to the PR #2585 to introduce CAM coupling with theta-l model. This PR uses one subroutine applycamforcing_tracers() to adjust tracers and to convert dynamics forcing (the crucial one is theta_dp forcing). The idea is that tracers update and recompute of forcing for theta_dp should be done at the same time. Now both stepon and standalone homme call this routine. To make it bfb with ftype2 we use a flag to interpret FQ as tendencies or adjustment. 

Restart routines have calls to endrun() to prevent restart in theta-l. This is not the best solution since restart is called by default at the end of the run, so, model with cam_target=theta-l crashes at the end if restart date is not pushed further. Restart for theta-l will be fixed later.

Also, ftype3 code is removed. Theta-c code (which is not tested) is now broken because code in set_thermostate and tests_finalize is disabled.

### CODE REVIEW

**Comment by oksanaguba** [comment]
For now homme standalone tests pass with 'strict' build (I ran 4, ftype4 and theta-test22, thetanhwet-TC), ERP_Ln9.ne4_ne4.FC5AV1C-L.anvil_intel pass, and whole model with theta-l cracshes at the 1st step in homme in reprosum. w is probably not set, maybe something else.

**Comment by oksanaguba** [comment]
This commit: homme standalone passes with strict.

**Comment by oksanaguba** [comment]
theta-l ran for ne4 for 5 days when restart point was pushed further.

**Comment by oksanaguba** [comment]
homme test suite passed (so moving code around did not affect bfb) and the same ERP test passes.

**Comment by oksanaguba** [comment]
tests for ftypes:

ftype0 IDENTICAL
```
/home/onguba/as/ftype0-branch-ne4/run
[onguba@b717 run]$ mpiexec -n 1 /home/ccsm-data/tools/cprnc -m ftype0-branch-ne4.cpl.r.0001-01-06-00000.nc ../../ftype0-master-ne4/run/ftype0-master-ne4.cpl.r.0001-01-06-00000.nc 
```
ftype1 DIFFERENT (as expected, not bfb, but is code correct?)
```
/home/onguba/as/ftype1-branch-ne4/run
mpiexec -n 1 /home/ccsm-data/tools/cprnc -m ftype1-branch-ne4.cpl.r.0001-01-06-00000.nc ../../ftype1-master-ne4/run/ftype1-master-ne4.cpl.r.0001-01-06-00000.nc
```

Script for these runs is below.

**Comment by oksanaguba** [comment]
```
#!/bin/bash -e
#The -e option means "if any pipeline ever ends with a non-zero ('error') exit status, terminate the script immediately". 

ftype=4 # or 0 or 4


#if [  1 -eq 2 ]; then

########################################## BRANCH
which=acme-b3
fol=ftype${ftype}-branch-ne4
${which}/cime/scripts/create_newcase -case $fol -res ne4_ne4 -mach anvil -compiler intel -compset FC5AV1C-L
cd $fol

./xmlchange JOB_WALLCLOCK_TIME=1:00:00
./xmlchange NTASKS=16
./xmlchange --file env_mach_pes.xml NTASKS_ESP=1
./xmlchange ROOTPE=0
./case.setup
uf=user_nl_cam
#first empty line
echo "" > $uf
#add your defs here
echo "se_ftype=${ftype}" >> $uf
./case.build
./xmlchange STOP_N=5 ; ./xmlchange STOP_OPTION=ndays ;
./case.submit
id=`grep -o case.run:\[0-9\]\* CaseStatus | sed -e 's/.*://' | tail -1`
echo $id
cp user_nl_cam usercam."${id}"
cd ${HOME}

#fi

########################################## MASTER
which=acme-master
fol=ftype${ftype}-master-ne4
${which}/cime/scripts/create_newcase -case $fol -res ne4_ne4 -mach anvil -compiler intel -compset FC5AV1C-L
cd $fol

./xmlchange JOB_WALLCLOCK_TIME=1:00:00
./xmlchange NTASKS=16
./xmlchange --file env_mach_pes.xml NTASKS_ESP=1
./xmlchange ROOTPE=0
./case.setup
uf=user_nl_cam
#first empty line
echo "" > $uf
#add your defs here
echo "se_ftype=${ftype}" >> $uf
./case.build
./xmlchange STOP_N=5 ; ./xmlchange STOP_OPTION=ndays ;
./case.submit
id=`grep -o case.run:\[0-9\]\* CaseStatus | sed -e 's/.*://' | tail -1`
echo $id
cp user_nl_cam usercam."${id}"
```

**Comment by oksanaguba** [comment]
There are still a few comment lines that can be deleted. One can test ftype=1 case by running ftype2 case with se_nsplit=1. Then they should be bfb (if dynamics update is the same).

File: components/cam/src/dynamics/se/stepon.F90 [hunk 1]
```diff
@@ -288,129 +300,54 @@ subroutine stepon_run2(phys_state, phys_tend, dyn_in, dyn_out )
...
 
+         do k=1,nlev
+            do j=1,np
+               do i=1,np       
                   ! force V, T, both timelevels
-                  dyn_in%elem(ie)%state%v(i,j,:,k,tl_f)= &
-                       dyn_in%elem(ie)%state%v(i,j,:,k,tl_f) +  &
-                       dtime*dyn_in%elem(ie)%derived%FM(i,j,:,k)
-                  
+                  do velcomp=1,2
```
Comment (oksanaguba) [inline components/cam/src/dynamics/se/stepon.F90]: For ftype1 for dynamics update, should I have called applycamforcing_dynamics from homme? This can be another pr which will include verification of ftype1 + maybe a new test for ftype1.

**Comment by oksanaguba** [comment]
This PR can be used for runs with nonhydro dycore as well. There are still details to figure out, like which pressure to use, but infrastructure is ok for some prelim runs. One question that came up was whether w_i on the bottom is init-ed correctly in EAM. There is a BC on w_i(nlevp) that computes it from grad of geopotential, but also w_i is set to 0 in read_inidat(). Code looks like this:
```
sub cam_initial()
  call initial_cond()
      call read_initial()
  ....
  call dyn_init2()
      call prim_init2()
          call model_init2()
              computes w_i(nlevp)
end cam_initial
```
and so w_i is init-ed correctly.

This PR has restart for theta broken. There is a trivial PR that fixes restart in EAM with theta model, but only for hydro. For nonhydro restart there will be another, a less trivial, PR.

**Comment by rljacob** [comment]
Is this PR ready to merge?

**Comment by oksanaguba** [comment]
yes

**Comment by oksanaguba** [comment]
This PR breaks bfb for ftype1 runs with cam, because it replaces tracers adjustment 
fq=dp * (FQ-Q); Qdp+=fq
with the same adjustment as in ftype0,
Qdp=dp * FQ.

**Comment by oksanaguba** [comment]
It appears that ftype1 is fine here. Ran these tests with nsplit1:
```
#!/bin/bash -e
#The -e option means "if any pipeline ever ends with a non-zero ('error') exit status, terminate the script immediately". 

#default for ne4 is
#ATM_NCPL=12

ftype=1 # or 2


#if [  1 -eq 2 ]; then

########################################## BRANCH
which=acme-b3
fol=ftype${ftype}-branch-ne4-nsplit1
${which}/cime/scripts/create_newcase -case $fol -res ne4_ne4 -mach anvil -compiler intel -compset FC5AV1C-L
cd $fol

./xmlchange JOB_WALLCLOCK_TIME=1:00:00
./xmlchange NTASKS=16
./xmlchange --file env_mach_pes.xml NTASKS_ESP=1
./xmlchange ROOTPE=0
./xmlchange ATM_NCPL=48
./case.setup
uf=user_nl_cam
#first empty line
echo "" > $uf
#add your defs here
echo "se_ftype=${ftype}" >> $uf
echo "se_nsplit=1" >> $uf

./case.build
./xmlchange STOP_N=5 ; ./xmlchange STOP_OPTION=ndays ;
./case.submit
id=`grep -o case.run:\[0-9\]\* CaseStatus | sed -e 's/.*://' | tail -1`
echo $id
cp user_nl_cam usercam."${id}"
cd ${HOME}

#fi


ftype=2

#if [  1 -eq 2 ]; then

########################################## BRANCH
which=acme-b3
fol=ftype${ftype}-branch-ne4-nsplit1
${which}/cime/scripts/create_newcase -case $fol -res ne4_ne4 -mach anvil -compiler intel -compset FC5AV1C-L
cd $fol

./xmlchange JOB_WALLCLOCK_TIME=1:00:00
./xmlchange NTASKS=16
./xmlchange --file env_mach_pes.xml NTASKS_ESP=1
./xmlchange ROOTPE=0
./xmlchange ATM_NCPL=48
./case.setup
uf=user_nl_cam
#first empty line
echo "" > $uf
#add your defs here
echo "se_ftype=${ftype}" >> $uf
echo "se_nsplit=1" >> $uf

./case.build
./xmlchange STOP_N=5 ; ./xmlchange STOP_OPTION=ndays ;
./case.submit
id=`grep -o case.run:\[0-9\]\* CaseStatus | sed -e 's/.*://' | tail -1`
echo $id
cp user_nl_cam usercam."${id}"
cd ${HOME}

#fi
```
and got identical files.
```
[onguba@b589 run]$ pwd
/home/onguba/as/ftype1-branch-ne4-nsplit1/run
[onguba@b589 run]$ mpiexec -n 1 /home/ccsm-data/tools/cprnc -m ftype1-branch-ne4-nsplit1.cpl.r.0001-01-06-00000.nc ../../ftype2-branch-ne4-nsplit1/run/ftype2-branch-ne4-nsplit1.cpl.r.0001-01-06-00000.nc
[onguba@b589 run]$ mpiexec -n 1 /home/ccsm-data/tools/cprnc -m ftype1-branch-ne4-nsplit1.cam.r.0001-01-06-00000.nc ../../ftype2-branch-ne4-nsplit1/run/ftype2-branch-ne4-nsplit1.cam.r.0001-01-06-00000.nc
```

**Comment by oksanaguba** [comment]
With ref states on,
```
diff --git a/components/homme/src/theta-l/prim_advance_mod.F90 b/components/homme/src/theta-l/prim_advance_mod.F90
index 4c80e62..93ddeb6 100644
--- a/components/homme/src/theta-l/prim_advance_mod.F90
+++ b/components/homme/src/theta-l/prim_advance_mod.F90
@@ -905,12 +905,12 @@ contains
      temp(:,:,:)=theta_ref(:,:,:,ie)*dp_ref(:,:,:,ie) 
      call get_phinh(hvcoord,elem(ie)%state%phis,&
           temp(:,:,:),dp_ref(:,:,:,ie),phi_ref(:,:,:,ie))
-#ifdef CAM
-     ! no reference state, for testing
-     theta_ref(:,:,:,ie)=0
-     phi_ref(:,:,:,ie)=0
-     dp_ref(:,:,:,ie)=0
-#endif
+!#ifdef CAM
+!     ! no reference state, for testing
+!     theta_ref(:,:,:,ie)=0
+!     phi_ref(:,:,:,ie)=0
+!     dp_ref(:,:,:,ie)=0
+!#endif
```
6 year run keeps crashing at day 476 in clubb, not in homme. Does this mean there is a bug in this PR?
```
[619]  xm_wpxp band solver: singular matrix
[619]  ERROR: lapack_wrap.F90: Singular matrix detected in band_solve
[619] Image              PC                Routine            Line        Source
[619] e3sm.exe           0000000002649FEB  Unknown               Unknown  Unknown
[619] e3sm.exe           00000000021FAC52  shr_abort_mod_mp_         114  shr_abort_mod.F90
[619] e3sm.exe           0000000001469F51  lapack_wrap_mp_ba         714  lapack_wrap.F90
[619] e3sm.exe           00000000013DCE15  advance_xm_wpxp_m        1878  advance_xm_wpxp_module.F90
[619] e3sm.exe           00000000013D7D47  advance_xm_wpxp_m         655  advance_xm_wpxp_module.F90
[619] e3sm.exe           00000000013BC885  advance_clubb_cor        1906  advance_clubb_core_module.F90
[619] e3sm.exe           0000000000FEE697  clubb_intr_mp_clu        1907  clubb_intr.F90
[619] e3sm.exe           00000000005E5A69  physpkg_mp_tphysb        2477  physpkg.F90
[619] e3sm.exe           00000000005DB6B9  physpkg_mp_phys_r        1028  physpkg.F90
[619] e3sm.exe           00000000004F2D11  cam_comp_mp_cam_r         250  cam_comp.F90
```

**Comment by oksanaguba** [comment]
There seems to be a bug in forcing conversion: for homme version, in theta calculations dp(n+1) is used, but in cam adjustment dp(n) is used because of extra variable.

**Comment by rljacob** [comment]
What is the status of this PR?

**Comment by oksanaguba** [comment]
@rljacob waiting for review.

**Comment by rljacob** [comment]
@mt5555 is this ancient PR read or not?

**Comment by mt5555** [comment]
merged to next
